File: MATTHEW/leetcode/AIcoding/5.Qwen-3/erlang_solutions/890-find-and-replace-pattern.erl
-module(solution).
-export([find_and_replace_pattern/2]).

find_and_replace_pattern(Pattern, Words) ->
    lists:filter(fun(W) -> is_match(Pattern, W) end, Words).

is_match(P1, P2) ->
    maps:size(maps:from_list(associations(P1, P2))) == 2.

associations([], []) ->
    [];
associations([H1|T1], [H2|T2]) ->
    [{H1, H2} | associations(T1, T2)].

File: leetcode/AIcoding/3.DeepSeek-v3/rust_solutions/890-find-and-replace-pattern.rs
impl Solution {
    pub fn find_and_replace_pattern(words: Vec<String>, pattern: String) -> Vec<String> {
        let pattern = pattern.as_bytes();
        words.into_iter().filter(|word| {
            let word = word.as_bytes();
            if word.len() != pattern.len() {
                return false;
            }
            let mut word_to_pat = [0; 26];
            let mut pat_to_word = [0; 26];
            for i in 0..word.len() {
                let w = (word[i] - b'a') as usize;
                let p = (pattern[i] - b'a') as usize;
                if word_to_pat[w] == 0 && pat_to_word[p] == 0 {
                    word_to_pat[w] = p + 1;
                    pat_to_word[p] = w + 1;
                } else if word_to_pat[w] != p + 1 || pat_to_word[p] != w + 1 {
                    return false;
                }
            }
            true
        }).collect()
    }
}